lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -626,230 +626,68 @@ This is problematic for linalg ops for a few reasons:
-          Value kSizeInt = castIndexToInt64(b, loc, kSize);
-          Value kSizeF = b.create<arith::SIToFPOp>(loc, elementType, kSizeInt);
-          // accumulate out2 to res = args[1]
-          Value out2 = b.create<arith::AddFOp>(loc, res, out1);
-          b.create<linalg::YieldOp>(loc, ValueRange({out2, kSizeF}));
-        });
-
-    // make a linalg generic to divide each element by the corresponding
-    // Kernel Width. This step is only necessary for avg pooling.
-    SmallVector<AffineMap> indexingMaps1 =
-        AffineMap::inferFromExprList({kSizeTensorExprs, outputExprs});
-    SmallVector<utils::IteratorType> iteratorTypes1(
-        3, utils::IteratorType::parallel);
-    auto output = rewriter.create<linalg::GenericOp>(
-        loc, /*resultTensorTypes=*/initOutput.getType(),
-        /*inputs=*/sumPool.getResultTensors()[1],
-        /*outputs=*/sumPool.getResultTensors()[0],
-        /*indexingMaps=*/indexingMaps1,
-        /*iteratorTypes=*/iteratorTypes1,
-        [&](OpBuilder &b, Location loc, ValueRange args) {
-          Value q = b.create<arith::DivFOp>(loc, args[1], args[0]);
-          b.create<linalg::YieldOp>(loc, q);
-        });
-
-    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, outputType,
-                                                output.getResultTensors());
-    return success();
-  }
-};
-} // namespace
-
-// The logic for this conversion is similar to the AdaptiveAvgPool1dOp
-// conversion. Before writing any more adaptive pooling conversions, the logic
-// in this should be off-loaded to a helper function, since each of the adaptive
-// ops are essentially the same with some minor tweaks. Instead of kSizeTensor,
-// we named the additional output of the linalg generic op auxTensor.
-// For max pooling, auxTensor holds the indices of max values, and for
-// avg pooling, the auxTensor will be kSizeTensor, used to later divide the
-// sum pool by the kernel size.
-namespace {
-class ConvertAtenAdaptiveMaxPool2dOp
-    : public OpConversionPattern<AtenAdaptiveMaxPool2dOp> {
-public:
-  using OpConversionPattern::OpConversionPattern;
-  LogicalResult
-  matchAndRewrite(AtenAdaptiveMaxPool2dOp op, OpAdaptor adaptor,
-                  ConversionPatternRewriter &rewriter) const override {
-
-    Location loc = op->getLoc();
-    const TypeConverter *typeConverter = getTypeConverter();
-
     // get rank of input (same as rank of output)
-    int64_t rank =
-        adaptor.getSelf().getType().cast<RankedTensorType>().getRank();
-    // input operand should be NCHW (i.e. rank 4)
-    if (rank != 4) {
-      return rewriter.notifyMatchFailure(op, "only supports input type NCHW");
+    int64_t rank = inputType.getRank();
+    // get number of non-spatial dims
+    int64_t nonSpatial;
+    if (rank == Dim + 2 || rank == Dim + 1) {
+      nonSpatial = rank - Dim;
+    } else {
+      return rewriter.notifyMatchFailure(
+          op, "only supports inputs with one or two non-spatial dims");
```

Comment:
Is this a constraint imposed by torch, or a temporary constraint? 

Simpler:

int64_t nonSpatial = rank - Dim;
if (nonSpatial != 1 && nonSpatial != 2) { ... }

Reply:
Ah that's a good point. This is already part of the shape information for these ops (from torch), so perhaps it should be worded differently or removed. I imagine if you tried to put something of rank 2 into an adaptive pool 3d op, it would cause errors in python before getting to this conversion. I'll test that when I get back to my PC, but it's likely that this check should just be removed, or at least replaced with a check for negative nonSpatial value, since that's the only thing that would cause genuine problems with this conversion even if the shape information is loosened later on.

Reply:
I guess there's not much harm is checking things multiple times